src/java/org/apache/cassandra/locator/AbstractReplicationStrategy.java

```diff
@@ -94,6 +94,16 @@ public ArrayList<InetAddress> getCachedEndpoints(Token t)
         return cachedEndpoints.get(t);
     }
 
+    public void clearCachedEndpoints()
+    {
+        synchronized (this)
+        {
+            logger.trace("clearing cached endpoints");
+            cachedEndpoints.clear();
+            lastInvalidatedVersion = tokenMetadata.getRingVersion();
+        }
+    }
+
```

Comment:
do we need this addition to be synchronized, like the method above it?

Reply:
need to think about it more. concurrency is tricky + semi-dangerous since we don't want to affect https://github.com/palantir/cassandra/pull/336 (which fixes race conditions that lead to corruption).

Relevant to the cache is this race fix, which we don't have. Though I don't really see how this would give us some ranges with 2 instead of 3 endpoints https://issues.apache.org/jira/browse/CASSANDRA-16673

Reply:
more generally, I wonder if we can/should just call TokenMetadata#invalidateCachedRings(), which will end up resetting the endpoint cache and incrementing the ring version

Reply:
Looking at where `cachedEndpoints` is queried, it's always safe for a `get` to return null, because that will force the code to re-compute. The double-check and lock from the method above seems to exist to prevent double-clearing the cache.

Reply:
oh, invalidateCachedRings() is for sure what we should be calling instead. Good find

Reply:
trying to think through how this could affect #336 -- all other invocations of invalidateCachedRings are called while we have taken out the token metadata read/write locks while this one will not. I think that is OK, b/c those instances are trying to synchronize updates to multiple in-memory maps that are used as a source of truth while here we're just forcing the next call to recompute the cached info and only touching this map. Anything else you can think of?

Reply:
I would think the lock changes in that PR are there because the topology is changing, and the cache is invalidated because the ring is changing.

Since we're just invalidating a cache I don't see why we'd need to lock.

The call we're making also just invalidates the cache, it doesn't clear the Map itself, so we don't have to worry about concurrent mutations of a collection.